### ISSUE TITLE
I want to know is marathon suport docker using gpu?

### ISSUE THREAD

**Issue by freepjf** [mesosphere/marathon#4928]
we want to use marathon to execut docker with gpu. is it suport?
can you give a sample example

**Comment by houshanren** [comment]
we used nvidia-docker,https://github.com/NVIDIA/nvidia-docker
but it not support using marathon with "gpus" config。

**Comment by nfnt** [comment]
If you need GPU resource isolation as provided by Mesos (see http://mesos.apache.org/documentation/latest/gpu-support), you'll have to use the Mesos containerizer for this. The Mesos containerizer [supports Docker images](http://mesos.apache.org/documentation/latest/docker-volume/).

**Comment by EMChamp** [comment]
I am also stuck on this issue. Currently I am trying to launch a docker instance that has access to GPU resources. 

AFAIK docker does not have built in support to access NVIDIA GPU resources, you must launch with nvidia-docker to use GPU resources inside a container. So it seems it is impossible to launch a container using the docker or mesos containerizer in Marathon that has access to the GPU resources. 

I launched a container with the following JSON, but after connecting to it, looks like it does not have access to nvidia-smi:

```
{
  "id": "nvidiatest1",
  "cmd": "python3 -m http.server 8080",
  "cpus": 1,
  "mem": 64.0,
  "instances": 1,
  "constraints": [["hostname", "CLUSTER", "stg-aacpgpu101.stg.jp.local"]],
  "container": {
    "type": "DOCKER",
    "docker": {
      "image": "nvidia/cuda:7.5-devel",
      "network": "BRIDGE"
    }
  }
```
}


Ideally it would be launched through the mesos containerizer (so that it can consume GPU resources), I've tried this JSON but it only runs the python server on the host:
```

{
  "id": "/nvidiatest3",
  "cmd": "python3 -m http.server 8080",
  "cpus": 1,
  "gpus": 1,
  "mem": 128,
  "disk": 0,
  "instances": 1,
  "container": {
        "type": "MESOS",
        "docker": {
            "image": "nvidia/cuda:7.5-devel",
            "credential": null
        }
    }
}
```

The only way I have been able to launch a container that will use GPU resources on Marathon is as a CMD argument. It seems to work as intended for the most part however when I scale down the service on Marathon it doesn't stop the container. 

```
{
  "id": "/nvidiatest2",
  "cmd": "docker run nvidia/cuda:7.5-devel python3 -m http.server 8080",
  "cpus": 1,
  "gpus": 1,
  "mem": 128,
  "disk": 0,
  "instances": 1
}

```
Basically I am still looking for a example of JSON you can send to mesos that will launch a container with GPU resources that exits/removes the containers successfully.

**Comment by jdef** [comment]
/nvidiatest3 looks like a good start. you can use CNI networks w/ the mesos
containerizer: define `ipAddress { name: "your-network" }`, where
`your-network` is the name of a CNI network. be sure to make necessary
configuration changes to your mesos servers to support CNI.

**Comment by c-yokoyama** [comment]
Now, I am trying the similar case and  used nvidia-docker like this.
But, it didn't launch a container well.
```
{
    "id": "nvidia-docker-test",
    "cmd": "nvidia-docker run -it -d nvidia/cuda",
    "cpus": 1,
    "mem": 512.0,
    "gpus": 1
}
```
I'd like to know a example of JSON using nvidia-docker command.

**Comment by EMChamp** [comment]
After doing some digging you can launch docker containers with GPU capabilities using privileged mode and passing the docker volume containing nvidia drivers as noted in the nvidia-docker wiki in the alternatives section:

https://github.com/NVIDIA/nvidia-docker/wiki/NVIDIA-driver

I had to add privileged mode to the command to give access to the devices (alternatively you can specify the devices you want to give the container access to). Currently I only got this to work with the Docker containerizer, there is a curl error with docker registry I am running into with the Mesos containerizer that I am trying to figure out.

JSON to start container:
{
  "id": "/marathon-gpu-test",
  "groups": [
    {
      "id": "agents",
      "apps": [
        {
          "id": "marathon-gpu-test",
	  "cmd": "python3 -m http.server 8080",
          "cpus": 1,
          "mem": 512.0,
          "instances": 1,
          "container": {
            "type": "DOCKER",
            "docker": {
              "image": "nvidia/cuda:7.5-devel",
              "network": "HOST",
	      "privileged": true,
	      "parameters": [
                {
                  "key": "publish",
                  "value": "10080:10080"
                },
                {
                  "key": "volume-driver",
                  "value": "nvidia-docker"
                },
                {
                  "key": "volume",
                  "value": "nvidia_driver_367.48:/usr/local/nvidia:ro"
                }
	      ]
            }
          },
          "constraints": [
            [
              "hostname",
              "LIKE",
              "fakeserver"
            ],
            [
              "hostname",
              "UNIQUE"
            ]
          ]
        }
      ]
    }
  ]
}

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-4421. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
